astropy/coordinates/baseframe.py

```diff
@@ -224,50 +236,111 @@ def __init__(self, *args, **kwargs):
             repr_kwargs = {}
             for nmkw, nmrep in self.representation_component_names.items():
                 if len(args) > 0:
-                    #first gather up positional args
+                    # first gather up positional args
                     repr_kwargs[nmrep] = args.pop(0)
                 elif nmkw in kwargs:
                     repr_kwargs[nmrep] = kwargs.pop(nmkw)
 
-            #special-case the Spherical->UnitSpherical if no `distance`
-            #TODO: possibly generalize this somehow?
+            # special-case the Spherical->UnitSpherical if no `distance`
+            # TODO: possibly generalize this somehow?
 
             if repr_kwargs:
                 if repr_kwargs.get('distance', True) is None:
                     del repr_kwargs['distance']
-                if (issubclass(self.representation, SphericalRepresentation) and
-                        'distance' not in repr_kwargs):
+
+                if (issubclass(self.representation, r.SphericalRepresentation)
+                        and 'distance' not in repr_kwargs):
                     representation = self.representation._unit_representation
                 else:
                     representation = self.representation
+
                 representation_data = representation(copy=copy, **repr_kwargs)
 
+            # Now we handle the Differential data:
+            # Get any differential data passed in to the frame initializer
+            # using keyword or positional arguments for the component names
+            diff_kwargs = {}
+            for nmkw, nmrep in self.differential_component_names.items():
+                if len(args) > 0:
+                    # first gather up positional args
+                    diff_kwargs[nmrep] = args.pop(0)
+                elif nmkw in kwargs:
+                    diff_kwargs[nmrep] = kwargs.pop(nmkw)
+
+            if diff_kwargs:
+                if (hasattr(self.differential_cls, '_unit_differential') and
+                        'd_distance' not in diff_kwargs):
+                    differential = self.differential_cls._unit_differential
+
+                elif len(diff_kwargs) == 1 and 'd_distance' in diff_kwargs:
+                    differential = r.RadialDifferential
+
+                else:
+                    differential = self.differential_cls
+                differential_data = differential(copy=copy, **diff_kwargs)
+
         if len(args) > 0:
             raise TypeError(
                 '{0}.__init__ had {1} remaining unhandled arguments'.format(
                     self.__class__.__name__, len(args)))
 
+        if representation_data is None and differential_data is not None:
+            raise ValueError("Cannot pass in differential component data "
+                             "without positional (representation) data.")
+
         self._data = representation_data
+        if self._data and differential_data is not None:
```

Comment:
You just checked that if `differential_data is not None`, `representation_data` is defined. So, maybe
```
if differential_data:
    self._data = representation_data.with_differentials({'s': differential_data})
else:
   self._dat a= representation_data  # possibly None.
```